src/System.Private.CoreLib/src/Internal/Runtime/Augments/RuntimeThread.Unix.cs

```diff
@@ -2,62 +2,133 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using Microsoft.Win32.SafeHandles;
 using System;
 using System.Diagnostics;
+using System.Runtime.InteropServices;
 using System.Threading;
 
 namespace Internal.Runtime.Augments
 {
     public sealed partial class RuntimeThread
     {
-        private readonly WaitSubsystem.ThreadWaitInfo _waitInfo;
+        private static bool s_setThreadExitCallback;
 
-        private void PlatformSpecificInitialize() { }
+        // Event signaling that the thread has stopped
+        private ManualResetEvent _stopped;
 
-        private void PlatformSpecificInitializeExistingThread() { }
+        private readonly WaitSubsystem.ThreadWaitInfo _waitInfo;
 
         internal WaitSubsystem.ThreadWaitInfo WaitInfo => _waitInfo;
 
-        private static ThreadPriority GetPriority() { return ThreadPriority.Normal; }
+        private void PlatformSpecificInitialize()
+        {
+            // Race condition is OK here since we set to the same value
+            if (!s_setThreadExitCallback)
+            {
+                Interop.Sys.RuntimeThread_SetThreadExitCallback(AddrofIntrinsics.AddrOf<Action>(OnThreadExit));
+                s_setThreadExitCallback = true;
+            }
+        }
 
-        private static bool SetPriority(ThreadPriority priority) { return true; }
+        private void PlatformSpecificInitializeExistingThread()
+        {
+            _stopped = new ManualResetEvent(false);
+        }
 
-        private bool HasFinishedExecution()
+        /// <summary>
+        /// Returns true if the thread is started or being started in StartInternal.
+        /// </summary>
+        private bool HasStarted()
         {
-            // TODO: Return true if the thread has finished execution
-            return false;
+            return _stopped != null;
         }
 
-        private bool JoinCore(int millisecondsTimeout)
+        private ThreadPriority GetPriorityLive()
         {
-            // TODO: Join the thread
-            return false;
+            return ThreadPriority.Normal;
         }
 
-        private void StartCore(object parameter)
+        private bool SetPriorityLive(ThreadPriority priority)
         {
-            // TODO: Start a new thread using _threadStart and parameter
+            return true;
         }
 
-        public ApartmentState GetApartmentState()
+        [NativeCallable]
+        private static void OnThreadExit()
```

Comment:
@AntonLapounov, @jkotas, @sergiy-k we have to be really careful to not to use any `thread_local` object in the call chain of the OnThreadExit. Since we use a thread_local object as a mechanism to call the PalDetachThread on Unix and that object creation is triggered in the PalAttachThread, all other `thread_local` objects that runtime creates would be already destroyed when OnThreadExit is called. 
At the moment, we use `thread_local` c++ feature only for this thread shutdown notification and we have other thread local variables declared using the __thread keyword that are destroyed only after all thread_local objects are destroyed. But we may decide later to switch all thread local variables to `thread_local` and then we might start hitting problems.

I wonder, could we instead pass the pthread_create a managed wrapper that would wrap a call to the actual thread function in `try` / `finally`, calling OnThreadExit in `finally`? Then we would stay in the confines of the fully alive managed thread. Or do we also care about calling OnThreadExit for foreign threads that got attached to runtime, e.g. by reverse pinvoke from a foreign thread?